### PR TITLE
Chris/feature/dashboard

### PR DESCRIPTION
**Pull request by flumonion**

### CODE REVIEW

**Comment by TomRadford** [comment]
flagging if this hover-over menu should do anything:
image

Can resolve if it doesn't have to do with this PR

**Comment by TomRadford** [comment]
Something that doesn't make sense to me from a UI perspective is these two dropdowns (the one is open with four options):
image

Changing one of them changes the other which causes the whole UI to revert to Skeleton. 
I believe that this dropdown could be placed above or somewhere where it isn't replicated ie:
image

In Dashboard.svelte this could be:
```
  <div class="mt-7">
    <div class="mb-2 mt-10 flex w-full justify-center self-center">
      <Dropdown
        options={getGraphEntities()}
        bind:value={entity}
        className="w-24 mr-[27px] "
        inputClass="font-bold text-small text-neutral-700 px-3 py-0 border-neutral-200 h-[24px] rounded-[5px]"
      />
    </div>
    <div class="grid gap-6 lg:grid-cols-1 xl:grid-cols-2">
      <DashboardInfoError bind:entity logs={logsIngress} direction="ingress" {isLoading} {config} />
      <DashboardInfoError bind:entity logs={logsEgress} direction="egress" {isLoading} {config} />
    </div>
  </div>
```

File: package.json [hunk 1]
```diff
@@ -83,6 +83,8 @@
     "@tanstack/svelte-query": "^4.33.0",
     "chart.js": "^4.3.0",
     "clsx": "^2.0.0",
+    "cron-parser": "^4.9.0",
```
[the remark this answers is not in the document]
Comment (flumonion) [reply package.json:86]: it is being used `DashboardSyncPull`

File: s2s.dic
[the remark this answers is not in the document]
Comment (flumonion) [reply s2s.dic:1]: i use it to remove errors in my ide, if that does not bother you, you can leave it.

File: src/components/Dashboard/DashboardSync.svelte [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    return getSourceSyncMode(source) === 'pull';
+  };
+</script>
+
+{#if sources.length > 0}
+  <div class="w-full rounded bg-white text-regular shadow">
+    <div class="flex flex-wrap items-center justify-between gap-6 p-6 text-neutral-800">
+      <!-- Title-->
+      {#if isLoading}
+        <Skeleton class="h-8 w-36" shimmer={true} />
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardSync.svelte:42]: can you give an example of the shift?

File: src/components/Dashboard/Dashboard.svelte [hunk 3]
```diff
@@ -1,170 +1,100 @@
...
+    if (!source) {
+      source = getSource(config);
+    }
+  }
 </script>
 
 <div>
   <div>
-    <DashboardHeader />
+    <DashboardHeader {isLoading} bind:range />
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte:79]: removed.

File: src/components/Dashboard/Dashboard.svelte [hunk 4]
```diff
@@ -1,170 +1,100 @@
...
-      },
-      {
-        label: 'Errors',
-        value: 32,
-        className: 'text-brand-action'
-      }
-    ]
-  };
+  // Queries for fetching logs and config
+  $: queryConfig = createConfigQuery();
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 5]
```diff
@@ -1,170 +1,100 @@
...
+  $: queryLogsIngress = createLogsInfoErrorQuery(range, entity, 'ingress');
+  $: queryLogsEgress = createLogsInfoErrorQuery(range, entity, 'egress');
+  $: queryLogsSourceSync = createLogsSourceSyncQuery();
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 6]
```diff
@@ -1,170 +1,100 @@
...
-      },
-      {
-        label: 'Errors',
-        value: 12,
-        className: 'text-brand-action'
-      }
-    ]
-  };
+  // check if trade channel and fulfillment service exist
+  $: tradeChannel = $queryConfig.data && hasTradeChannel($queryConfig.data);
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 7]
```diff
@@ -1,170 +1,100 @@
...
+  $: fulfillmentService = $queryConfig.data && hasFulfillmentService($queryConfig.data);
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 8]
```diff
@@ -1,170 +1,100 @@
...
-        label: 'Errors',
-        value: 32,
-        className: 'text-brand-action'
-      }
-    ]
-  };
+  // Get logs for both ingress and egress
+  let logsIngress: InfoErrorResponse;
+  $: if ($queryLogsIngress.data) {
+    logsIngress = $queryLogsIngress.data;
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 9]
```diff
@@ -1,170 +1,100 @@
...
+  }
+  let logsEgress: InfoErrorResponse;
+  $: if ($queryLogsEgress.data) {
+    logsEgress = $queryLogsEgress.data;
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 10]
```diff
@@ -1,170 +1,100 @@
...
-        backgroundColor: 'rgba(53, 162, 235, 0.5)',
-        pointBackgroundColor: 'white',
-        borderWidth: 1
-      }
-    ]
-  };
+  // Get source sync logs
+  let sourceSync: SourceSyncResponse;
+  $: if ($queryLogsSourceSync.data) {
+    sourceSync = $queryLogsSourceSync.data;
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 11]
```diff
@@ -1,170 +1,100 @@
...
+    !sourceSync ||
+    !$queryLogsEgress.isFetchedAfterMount ||
+    !$queryLogsIngress.isFetchedAfterMount;
+
+  // set config and source from config
+  let config: Config;
+  $: if ($queryConfig.data) {
+    config = $queryConfig.data;
+    if (!source) {
+      source = getSource(config);
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/a697f88690d67428aac940615a4d4534d0a244b4

File: src/components/Dashboard/Dashboard.svelte [hunk 12]
```diff
@@ -1,170 +1,100 @@
...
+    $queryConfig.isLoading ||
+    !$queryConfig.data ||
+    !logsIngress ||
+    !logsEgress ||
+    !sourceSync ||
+    !$queryLogsEgress.isFetchedAfterMount ||
+    !$queryLogsIngress.isFetchedAfterMount;
+
+  // set config and source from config
+  let config: Config;
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/Dashboard.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/7a8eb3d05a41664032e16358fa7efd489103a615

File: src/components/Dashboard/DashboardInfoBlock.svelte [hunk 13]
```diff
@@ -1,58 +1,80 @@
...
+
+  let incomingErrors = 0;
+  let incomingInfo = 0;
+  let outgoingErrors = 0;
+  let outgoingInfo = 0;
+
+  // set incoming values
+  $: if (logsIngress) {
+    let errors = logsIngress.aggregations?.range?.buckets?.errors?.entities?.buckets;
+    let info = logsIngress.aggregations?.range?.buckets?.info?.entities?.buckets;
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardInfoBlock.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/c14c0476e2feae92be228111296fe90d0037ec66

File: src/components/Dashboard/DashboardInfoError.svelte [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    {/if}
+  </div>
+
+  <div class="mt-4">
+    <DashboardInfoErrorList {isLoading} {hits} {config} />
+  </div>
+
+  {#if !isLoading}
+    <div>
+      <Btn on:click={goToLogs} class="mt-6" size="small" variant="gray-outline">View Logs</Btn>
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardInfoError.svelte:73]: I did not want to have to reinvent the design of the button, is that not a good trade off?

File: src/components/Dashboard/DashboardInfoError.svelte [hunk 15]
```diff
@@ -0,0 +1,76 @@
+<script lang="ts">
+  import Icon from '../Icons/Icon.svelte';
+  import IconDatabase from '../Icons/IconDatabase.svelte';
+  import Dropdown from '../input/Dropdown.svelte';
+  import IngressChart from '@/components/Dashboard/DashboardInfoErrorChart.svelte';
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardInfoError.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/35574cd3d72ec4e53f7f00470b55b67e81003795

File: src/components/Dashboard/DashboardInfoErrorChart.svelte [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            backgroundColor: 'rgba(255, 99, 132, 0.5)',
+            pointBackgroundColor: 'white',
+            borderWidth: 1
+          }
+        ],
+        labels: []
+      }
+    };
+    let errors = esData.aggregations?.range?.buckets?.errors?.entities?.buckets;
+    let info = esData.aggregations?.range?.buckets?.info?.entities?.buckets;
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardInfoErrorChart.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/2a22408bbf25fa473578f965befd99b0da2938c6

File: src/components/Dashboard/DashboardInfoErrorList.svelte
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardInfoErrorList.svelte:1]: not following this comment, can you link to the code?

File: src/components/Dashboard/DashboardHeader.svelte [hunk 17]
```diff
@@ -1,20 +1,19 @@
 <script lang="ts">
-  import Dropdown from '../input/Dropdown.svelte';
-  import DashboardTimeFrameFilter from './DashboardTimeFrameFilter.svelte';
+  import DashboardTimeFrameFilter from './DashboardRangeSelector.svelte';
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardHeader.svelte]: https://github.com/stock2shop/console_v2/pull/368/commits/8e623e7672c2d8771ebcacfc55c435c4aa4327eb

File: src/components/Dashboard/DashboardRangeSelector.svelte [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  export const keys = getRangeKeys();
+  export let range: RangeKey;
+</script>
+
+<div
+  class="flex h-[32px] items-center rounded border border-neutral-200 p-[2px] font-bold text-neutral-700"
+>
+  {#each keys as key}
+    <div
+      on:click={() => (range = key)}
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardRangeSelector.svelte:14]: https://github.com/stock2shop/console_v2/pull/368/commits/3431466597509f85696cb47767e828b77648e9a8

File: src/components/Dashboard/DashboardSync.svelte [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+  const isPush = (source: Source): boolean => {
+    return getSourceSyncMode(source) === 'push';
+  };
+  const isPull = (source: Source): boolean => {
+    return getSourceSyncMode(source) === 'pull';
+  };
+</script>
+
+{#if sources.length > 0}
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardSync.svelte:37]: yes, very low...
You cannot use the system without a source.

File: src/components/Dashboard/DashboardSyncPull.svelte [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+        {:else}
+          Never or older than a week
+        {/if}
+      </span>
+    </div>
+  {/if}
+
+  <!-- Sync Button-->
+  <button
+    on:click={goToLogs}
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardSyncPull.svelte:111]: but then we need to restyle the href?

File: src/components/Dashboard/DashboardSyncPush.svelte
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardSyncPush.svelte:1]: see response

File: src/lib/date-utils.ts [hunk 21]
```diff
@@ -33,3 +33,24 @@ export const dateWithOrdinalDay = (date: Date): string => {
...
 };
+
+export const getLocalDate = (isoDateStringWithoutZone: string): Date => {
+  const str = `${isoDateStringWithoutZone} UTC`;
+  return new Date(str);
+};
+
+export const getShortDay = (isoDateStringWithoutZone: string): string => {
+  const date = getLocalDate(isoDateStringWithoutZone);
+  const short = new Intl.DateTimeFormat('en', {
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/lib/date-utils.ts:50]: bug fix here.

File: src/models/config/config.ts [hunk 22]
```diff
@@ -79,3 +79,105 @@ export interface Flag {
   description: string;
   code: string;
 }
+
+/**
+ * Check if there is a trade channel
+ * @param config
```
[the remark this answers is not in the document]
Comment (flumonion) [reply src/models/config/config.ts:85]: my ide added this one in, have removed it.
https://github.com/stock2shop/console_v2/pull/368/commits/04b55c1c4d9b6c3a9687fd78cc478d22dc83b766

File: src/components/Dashboard/DashboardSync.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/components/Dashboard/DashboardSync.svelte:42]: It was when we do not know the sources count for:
image
so we cant determine the height of the skeleton

File: src/stories/test-data/repos/logs-repo.ts [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+export default class LogsRepo {
+  basePath = 'logs';
+
+  async getInfoError(
+    range: RangeKey,
+    entity: EntityKey,
+    direction: DirectionKey
+  ): Promise<InfoErrorResponse> {
+    // mock es response data gets imported and responded here
+    return undefined as unknown as InfoErrorResponse;
```
Comment (TomRadford) [inline src/stories/test-data/repos/logs-repo.ts:18]: @flumonion starting point to return mock data

File: src/components/Dashboard/DashboardInfoError.svelte (hunk 14 above)
[the remark this answers is not in the document]
Comment (TomRadford) [reply src/components/Dashboard/DashboardInfoError.svelte:73]: you don't need to :) - just do:
```
<a href='/logs'>
<Btn class="mt-6" size="small" variant="gray-outline">View Logs</Btn>
</a>
```

**Review by TomRadford — CHANGES_REQUESTED** [review]

File: src/stories/test-data/repos/logs-repo.ts (hunk 23 above)
Comment (TomRadford) [reply src/stories/test-data/repos/logs-repo.ts:18]: you can ignore this (just leave it for now) if you remove `src/stories/routes/98-Dashboard.stories.svelte`

File: src/components/Dashboard/DashboardSync.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (flumonion) [reply src/components/Dashboard/DashboardSync.svelte:42]: ja, so that may be a rare case... prepared to leave this for now?

**Review by TomRadford — APPROVED** [review]

**Pull request merged**